### DISCUSSION TITLE
Parametrise SQL queries with templated variables via BigQueryInsertJobOperator

### DISCUSSION THREAD

**Discussion by olivierbenard** [apache/airflow#23526] [Q&A] [answered]
Hi!

**TL;DR:** `BigQueryInsertJobOperator` cannot parametrise all the parameters of a parametrised SQL query. For instance, Google BigQuery named variables cannot be structural parts of the query - e.g. `projectID`. Is there a global solution though, that would allow parametrising all variables (templated, not-templated, structural components) at one central place (see **option 2** example) without having to define variables on top of each SQL queries (see **option 1**)?

**More context:**

Let’s say you have the following `my_query.sql`, containing the following parametrised query:

```
SELECT * 
FROM `{{src_project}}.src-dataset.src-table`
WHERE DATE(_PARTITIONTIME) = '{{partitiontime}}';
```

This SQL is parametrised with 2 templated variables:
-	`src_project` = `if env.prod ? src-project : src-project-dev`
-	`partitiontime` = `"{{ next_ds }}"`

A first option - which I will call **option 1** - would be to add the following lines in each SQL files you have, having the variables defined on top of each SQL queries:

```
{% set src_project = var.value.src_project %}
{% set partitiontime = next_ds %}

SELECT * 
FROM `{{src_project}}.src-dataset.src-table`
WHERE DATE(_PARTITIONTIME) = '{{partitiontime}}';
```

This solution works but is not ideal.

I would like to have a more high-level organization, with the variables defined in one central place and not having these redundancies lying everywhere in the SQL files. Imagine the variables change, I would then have to change it everywhere. Even with a good IDE, something feels wrong with that premise.

Instead, I would like to pass on the variables directly at the job level - i.e. via the `BigQueryInsertJobOperator`’s attributes - in one central place, once and for all.

My ideal solution - which I will call **option 2** - would be:

```
# see https://github.com/apache/airflow/issues/22859
# this in order to template_searchpath .sql files.
class BigQueryTemplateFileOperator(BigQueryInsertJobOperator):
    template_ext = (".json", ".sql")

execute_query = BigQueryTemplateFileOperator(
            …,
            configuration={
                "query": {
                    "query": "my_query.sql",
                    "queryParameters": [
                        {
                            "name": "src_project",
                            "parameterType": {"type": "STRING"},
                            "parameterValue": {"value": "{{ var.value.src_project }}"},
                        },
                        {
                            "name": "partitiontime",
                            "parameterType": {"type": "STRING"},
                            "parameterValue": {"value": "{{ next_ds }}"},
                        },
                    ],
                    …,
                    "destinationTable": {
                        "projectId": "{{var.value.dest_project}}",
                        "datasetId": "dest-dataset",
                        "tableId": "dest-table",
                    },
                }
            },
        )
```

And having a SQL query looking more like:

```
SELECT * 
FROM @src_project.src-dataset.src-table`
WHERE DATE(_PARTITIONTIME) = @partitiontime;
```

However, this **option 2** does not work. As stated in the [documentation](https://cloud.google.com/bigquery/docs/parameterized-queries), GBQ Named Parameters cannot be used in the structural parts of the query – e.g. to define `projectID`, `datasetID` or `tableID`:

_Parameters cannot be used as substitutes for identifiers, column names, table names, or other parts of the query._

Other options might exist, but they have their own limitations:

-	`user_defined_macro` does not handle templated fields and introduce implicit knowledge – i.e. you have to know that variables defined there parametrise the SQL queries, this is not explicitly mentioned as they could be used anywhere else for anything else.
-	`BigQueryInsertJobOperator.configuration.params` does not handle templated fields.
-	`BigQueryInsertJobOperator.configuration.queryParameters` and GBQ named variables cannot be used to configure structural part of the SQL query – e.g. `@projectID`.`@datasetID`.`@tableID` or `@hash(column) as column_hashed`.

Thus, my question:

Is there a way to get closer from **option 2**? i.e. having all the parameters (templated, not-templated, structural components) defined at the `BigQueryInsertJobOperator` level, moving from setting the variables in each individual SQL files (as we have in the **option 1** example)?

**Note:** Besides named parameters does not improve performance compare to hard-encoded values - as the query and the job configuration containing the query parameters are both passed at the same time to BigQuery service - they however prevent SQL injection attacks. Downside is that they won't be rendered by Airflow UI's rendered template. This is why I would preferably like to use them, but I am also open to any other suggestions.

Thank you very much for your consideration!
Olivier

**Comment by potiuk** [comment, marked as answer]
I think this is something that you could do with your own (not airflow) JINJA of the SQL in your operator. This should be as simple as passig all the parameters as context to your own JINJA rendering. I don't think it could be generalized enough for any operator but if you just want to hande it in BigQueryInsertJobOperator it should be as simple as


```
class MyTemplatedOperator(BigQueryInsertJobOperator):
     def execute(context, context):
        self.ć()
        super().execute(contexT)

   self.parse_sql_through_jinja_with():
        internal_jinja_context = { "params": configuration } 
       self.sql = jinjafy(self.sql, internal_jinja_context)
```
